File: FitoGrano/Classes/clsGetOrganizationStructure.vb
Public Class clsOrganizationStructure
    Dim ImageCol As New DevExpress.Utils.ImageCollection    ' Коллекция иконок
    Dim ndList As New List(Of ConditionNodeList)            ' Коллекция состояний нодов
    Dim NodeList As ConditionNodeList                       ' Запись для добавления в коллекцию

    Public SelectionIdOld As Integer                        ' Ид предыдущей выбранной выбранной записи
    Public SelectionId As Integer = 99999                   ' Ид выбранной записи
    Public sender As DevExpress.XtraTreeList.TreeList       ' Элемент TreeList
    Public TableName As String = "OrganizationStructure"    ' Имя таблицы в ДС

    ' коллекции состяния нодов
    Private Structure ConditionNodeList
        Dim Id As Integer           ' Ид нода
        Dim Expanded As Boolean     ' Состояноие нода
    End Structure

    ' Заполняем коллекцию
    Private Sub Filling_ndList(Optional node As DevExpress.XtraTreeList.Nodes.TreeListNode = Nothing,
                               Optional IsFirstRun As Boolean = True)
        If IsFirstRun Then ndList.Clear()
        Select Case IsFirstRun
            Case True ' Первый запуск
                For Each nd As DevExpress.XtraTreeList.Nodes.TreeListNode In sender.Nodes
                    NodeList.Id = nd("Id")
                    NodeList.Expanded = nd.Expanded
                    ndList.Add(NodeList)
                    If nd.HasChildren Then Filling_ndList(nd, False)
                Next

            Case Else ' рекурсия
                For Each nd As DevExpress.XtraTreeList.Nodes.TreeListNode In node.Nodes
                    NodeList.Id = nd("Id")
                    NodeList.Expanded = nd.Expanded
                    ndList.Add(NodeList)
                    If nd.HasChildren Then Filling_ndList(nd, False)
                Next
        End Select
    End Sub

    Public Sub GetOrganizationStructure(Optional ExpandedRoot As Integer = 1,
                                        Optional ExpandedFilials As Integer = 1,
                                        Optional ExpandedGroups As Integer = 0,
                                        Optional ExpandedRoleJob As Integer = 0,
                                        Optional OnlyCurentFillial As String = "NULL"
                                        )

        If sender.Nodes.Count <> 0 Then Filling_ndList()
        EventChangedControl = False
        ImageCol.Clear()
        ImageCol.AddImage(My.Resources.company_16x16)
        ImageCol.AddImage(My.Resources.organization_16x16)
        ImageCol.AddImage(My.Resources.group_people_16x16)
        ImageCol.AddImage(My.Resources.performer_16x16)
        ImageCol.AddImage(My.Resources.current_16x16)

        SelectQueryData( TableName, "EXEC OrganizationStructureFunction 
                                @ExpandedRoot = " & ExpandedRoot & ", 
                                @ExpandedFilials = " & ExpandedFilials & ", 
                                @ExpandedGroups = " & ExpandedGroups & ", 
                                @ExpandedRoleJob = " & ExpandedRoleJob & ",
                                @OnlyCurentFillial = " & OnlyCurentFillial & ",
                                @Function = 1", "GetOrganizationStructure" )
        If sender.DataSource Is Nothing Then
            sender.ClearNodes()
            sender.DataSource = iDataSet.Tables(TableName)
            sender.KeyFieldName = "Id"
            sender.ParentFieldName = "ParentId"
            sender.StateImageList = ImageCol
            sender.PopulateColumns()
            HidenAllColumns_TreeList(sender)
            sender.Columns("Name").Visible = sender.Columns.Count <> 0
        End If
        EventChangedControl = True
        ' всем нодам присваиваем иконку
        For Each nd As DevExpress.XtraTreeList.Nodes.TreeListNode In sender.Nodes
            FillSelectImageIndex(nd)
        Next
        ' извлечение из коллекций состояний нодов
        For Each n In ndList
            If sender.FindNodeByKeyID(n.Id) IsNot Nothing Then
                sender.FindNodeByKeyID(n.Id).Expanded = n.Expanded
            End If
        Next
        sender.FocusedNode = sender.FindNodeByKeyID(SelectionId)
    End Sub

    ' цикл по всем дочерним узлам
    Private Sub FillSelectImageIndex(ByVal ParentNode As DevExpress.XtraTreeList.Nodes.TreeListNode)
        ParentNode.Expanded = ParentNode("IsExpanded")
        ParentNode.StateImageIndex = ParentNode.Level
        For Each nd As DevExpress.XtraTreeList.Nodes.TreeListNode In ParentNode.Nodes
            FillSelectImageIndex(nd)
        Next
    End Sub

    ' выделение текущего филиала
    Public Sub NodeCellStyle(sender As Object, e As DevExpress.XtraTreeList.GetCustomNodeCellStyleEventArgs)
        If EventChangedControl = False Then Exit Sub
        If e.Column.FieldName <> "Name" Then Exit Sub
        If e.Node("IsCurrent") Then
            e.Appearance.Font = New System.Drawing.Font("Tahoma", 8.25, FontStyle.Bold Or FontStyle.Underline)
            e.Node.StateImageIndex = 4
        End If
    End Sub
End Class

File: FitoGrano/CS/Classes/clsGetOrganizationStructure.cs
// VBConversions Note: VB project level imports
using System.Collections.Generic;
using System;
using System.Linq;
using System.Drawing;
using System.Diagnostics;
using Microsoft.VisualBasic;
using System.Xml.Linq;
using System.Windows.Forms;
// End of VB project level imports

using FitoGrano;

namespace FitoGrano
{
	public class clsOrganizationStructure
	{
		DevExpress.Utils.ImageCollection ImageCol = new DevExpress.Utils.ImageCollection(); // РљРѕР»Р»РµРєС†РёСЏ РёРєРѕРЅРѕРє
		List<ConditionNodeList> ndList = new List<ConditionNodeList>(); // РљРѕР»Р»РµРєС†РёСЏ СЃРѕСЃС‚РѕСЏРЅРёР№ РЅРѕРґРѕРІ
		ConditionNodeList NodeList; // Р—Р°РїРёСЃСЊ РґР»СЏ РґРѕР±Р°РІР»РµРЅРёСЏ РІ РєРѕР»Р»РµРєС†РёСЋ
		
		public int SelectionIdOld; // РРґ РїСЂРµРґС‹РґСѓС‰РµР№ РІС‹Р±СЂР°РЅРЅРѕР№ РІС‹Р±СЂР°РЅРЅРѕР№ Р·Р°РїРёСЃРё
		public int SelectionId = 99999; // РРґ РІС‹Р±СЂР°РЅРЅРѕР№ Р·Р°РїРёСЃРё
		public DevExpress.XtraTreeList.TreeList sender; // Р­Р»РµРјРµРЅС‚ TreeList
		public string TableName = "OrganizationStructure"; // РРјСЏ С‚Р°Р±Р»РёС†С‹ РІ Р”РЎ
		
		// РєРѕР»Р»РµРєС†РёРё СЃРѕСЃС‚СЏРЅРёСЏ РЅРѕРґРѕРІ
		public struct ConditionNodeList
		{
			public int Id; // РРґ РЅРѕРґР°
			public bool Expanded; // РЎРѕСЃС‚РѕСЏРЅРѕРёРµ РЅРѕРґР°
		}
		
		// Р—Р°РїРѕР»РЅСЏРµРј РєРѕР»Р»РµРєС†РёСЋ
		private void Filling_ndList(DevExpress.XtraTreeList.Nodes.TreeListNode node = null, bool IsFirstRun = true)
			{
			if (IsFirstRun)
			{
				ndList.Clear();
			}
			if (IsFirstRun == true) // РџРµСЂРІС‹Р№ Р·Р°РїСѓСЃРє
			{
				foreach (DevExpress.XtraTreeList.Nodes.TreeListNode nd in sender.Nodes)
				{
					NodeList.Id = System.Convert.ToInt32(nd["Id"]);
					NodeList.Expanded = nd.Expanded;
					ndList.Add(NodeList);
					if (nd.HasChildren)
					{
						Filling_ndList(nd, false);
					}
				}
			} // СЂРµРєСѓСЂСЃРёСЏ
			else
			{
				foreach (DevExpress.XtraTreeList.Nodes.TreeListNode nd in node.Nodes)
				{
					NodeList.Id = System.Convert.ToInt32(nd["Id"]);
					NodeList.Expanded = nd.Expanded;
					ndList.Add(NodeList);
					if (nd.HasChildren)
					{
						Filling_ndList(nd, false);
					}
				}
			}
		}
		
		public void GetOrganizationStructure(int ExpandedRoot = 1, int ExpandedFilials = 1, int ExpandedGroups = 0, int ExpandedRoleJob = 0, string OnlyCurentFillial = "NULL")
			{
			
			if (sender.Nodes.Count != 0)
			{
				Filling_ndList();
			}
			mdVariables.EventChangedControl = false;
			ImageCol.Clear();
			ImageCol.AddImage(global::My.Resources.Resources.company_16x16);
			ImageCol.AddImage(global::My.Resources.Resources.organization_16x16);
			ImageCol.AddImage(global::My.Resources.Resources.group_people_16x16);
			ImageCol.AddImage(global::My.Resources.Resources.performer_16x16);
			ImageCol.AddImage(global::My.Resources.Resources.current_16x16);
			
			
			sender.ClearNodes();
			sender.DataSource = mdQuery.iDataSet.Tables[TableName];
			sender.KeyFieldName = "Id";
			sender.ParentFieldName = "ParentId";
			sender.StateImageList = ImageCol;
			sender.PopulateColumns();
			mdHidenAllColumns.HidenAllColumns_TreeList(sender);
			sender.Columns["Name"].Visible = sender.Columns.Count != 0;
		}
		mdVariables.EventChangedControl = true;
		// РІСЃРµРј РЅРѕРґР°Рј РїСЂРёСЃРІР°РёРІР°РµРј РёРєРѕРЅРєСѓ
		foreach (DevExpress.XtraTreeList.Nodes.TreeListNode nd in sender.Nodes)
		{
			FillSelectImageIndex(nd);
		}
		// РёР·РІР»РµС‡РµРЅРёРµ РёР· РєРѕР»Р»РµРєС†РёР№ СЃРѕСЃС‚РѕСЏРЅРёР№ РЅРѕРґРѕРІ
		foreach (var n in ndList)
		{
			if (sender.FindNodeByKeyID(n.Id) != null)
			{
				sender.FindNodeByKeyID(n.Id).Expanded = System.Convert.ToBoolean(n.Expanded);
			}
		}
		sender.FocusedNode = sender.FindNodeByKeyID(SelectionId);
	}
	
	// С†РёРєР» РїРѕ РІСЃРµРј РґРѕС‡РµСЂРЅРёРј СѓР·Р»Р°Рј
	private void FillSelectImageIndex(DevExpress.XtraTreeList.Nodes.TreeListNode ParentNode)
	{
		ParentNode.Expanded = ParentNode["IsExpanded"];
		ParentNode.StateImageIndex = ParentNode.Level;
		foreach (DevExpress.XtraTreeList.Nodes.TreeListNode nd in ParentNode.Nodes)
		{
			FillSelectImageIndex(nd);
		}
	}
	
	// РІС‹РґРµР»РµРЅРёРµ С‚РµРєСѓС‰РµРіРѕ С„РёР»РёР°Р»Р°
	public void NodeCellStyle(object sender, DevExpress.XtraTreeList.GetCustomNodeCellStyleEventArgs e)
	{
		if (mdVariables.EventChangedControl == false)
		{
			return;
		}
		if (e.Column.FieldName != "Name")
		{
			return;
		}
		if (e.Node["IsCurrent"])
		{
			e.Appearance.Font = new System.Drawing.Font("Tahoma", (float) (8.25F), (System.Drawing.FontStyle) (FontStyle.Bold | FontStyle.Underline));
			e.Node.StateImageIndex = 4;
		}
	}
}

}
